client/admin/chat-channels.tsx

```diff
@@ -0,0 +1,242 @@
+  margin-top: 16px;
+  margin-bottom: 8px;
+`
+
+const StyledSearchInput = styled(SearchInput)`
+  width: 256px;
+`
+
+const SearchContainer = styled.div`
+  height: 88px;
+  margin: 16px 0;
+`
+
+const NoResults = styled.div`
+  ${subtitle1};
+
+  color: ${colorTextFaint};
+`
+
+const ErrorText = styled.div`
+  ${subtitle1};
+
+  color: ${colorError};
+`
+
+const StyledCarousel = styled(Carousel)`
+  max-width: 960px;
+`
+
+const ChannelThumbnail = styled(Card)`
+  width: 420px;
+  padding: 16px;
+
+  flex-shrink: 0;
+  display: flex;
+  flex-direction: row;
+  align-items: center;
+
+  &:not(:first-child) {
+    margin-left: 16px;
+  }
+`
+
+const StyledChannelIcon = styled(ChannelIcon)`
+  width: 56px;
+  height: auto;
+  flex-shrink: 0;
+`
+
+const ChannelInfoContainer = styled.div`
+  flex-grow: 1;
+  display: flex;
+  flex-direction: column;
+  margin: 0 8px;
+  overflow: hidden;
+`
+
+const ChannelInfoName = styled(Headline6)`
+  ${singleLine};
+`
+
+const ChannelHeadline = styled.div`
+  ${headline6};
+  margin-bottom: 8px;
+`
+
+const ChannelContainer = styled.div`
+  width: 100%;
+  overflow: hidden;
+
+  flex-grow: 1;
+  display: flex;
+`
+
+const StyledMessageList = styled(MessageList)`
+  flex-grow: 1;
+  max-width: 960px;
+  min-width: 320px;
+`
+
+export function AdminChatChannels() {
+  const dispatch = useAppDispatch()
+  const [channels, setChannels] = useState<ChannelInfo[]>()
+  const [currentPage, setCurrentPage] = useState(-1)
+  const [hasMoreChannels, setHasMoreChannels] = useState(true)
+
+  const [isLoadingMoreChannels, setIsLoadingMoreChannels] = useState(false)
+  const [searchError, setSearchError] = useState<Error>()
+  const [searchQuery, setSearchQuery] = useLocationSearchParam('q')
+  const abortControllerRef = useRef<AbortController>()
+
+  const carouselRef = useRef(null)
+  const debouncedSearchRef = useRef(
+    debounce((searchQuery: string) => {
+      // Just need to clear the search results here and let the carousel initiate the network
+      // request.
+      setSearchQuery(searchQuery)
+      setSearchError(undefined)
+      setChannels(undefined)
+      setCurrentPage(-1)
+      setHasMoreChannels(true)
+      if (carouselRef.current) {
+        // TODO(2Pac): Fix the types once the carousel is ported to TypeScript.
+        ;(carouselRef.current as any).reset()
+      }
+    }, 100),
+  )
+
+  const onSearchChange = useStableCallback((searchQuery: string) => {
+    debouncedSearchRef.current(searchQuery)
+  })
+
+  const onLoadMoreChannels = useStableCallback(() => {
+    setIsLoadingMoreChannels(true)
+    setCurrentPage(currentPage + 1)
+
+    abortControllerRef.current?.abort()
+    abortControllerRef.current = new AbortController()
+
+    fetchJson<ChannelInfo[]>(
+      apiUrl`admin/chat/?q=${searchQuery}&limit=${SEARCH_CHANNELS_LIMIT}&page=${currentPage + 1}`,
+      { signal: abortControllerRef.current.signal },
+    )
+      .then(data => {
+        setIsLoadingMoreChannels(false)
```

Comment:
Hmm, I'm not sure I quite understand the first part. Won't the `fetch` reject with an error if the signal is aborted and these won't get called? Also, can you clarify what you mean by storing a reference to AbortController outside of the ref.

Reply:
When promises resolve, the callbacks are called asynchronously. Thus you can have:
fetch() -> resolves
abort()
callback()

So you need to check that the AbortSignal has not become aborted since the fetch returned. Is this rare? Yes. Does it happen? Yes, so you need to handle it

As far as storing a reference outside of the ref, you need to store it in a local variable so you can access the same AbortSignal that you called `fetch` with (because in the case that it was aborted by a new fetch, your ref would be set to a new instance of `AbortController`, and thus have a new, non-aborted signal)